libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+      }
+    } else if constexpr (__is_move_only_function<remove_cvref_t<_Func>>::value) {
+      if (!__func) {
+        __storage_ = {};
+      } else {
+        auto& __data      = __storage_.__large_;
+        __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+        __data.__destroy_ = &_FuncWraps::__destroy;
+        __data.__heap_    = static_cast<std::byte*>(::operator new(sizeof(_Func), std::align_val_t(alignof(_Func))));
+        std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__heap_), std::forward<_Func>(__func));
+        __data.__status_ = _Store::_Status::_Heap;
+      }
+    } else {
+      __construct<_Func>(std::forward<_Func>(__func));
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI _Store::_Status __get_status() noexcept { return __storage_.__large_.__status_; }
+
+  template <class _Func, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(in_place_type_t<_Func>, _Args&&... __args) {
+    static_assert(is_same_v<decay_t<_Func>, _Func>);
+    __construct<_Func>(std::forward<_Args>(__args)...);
+  }
+
+  template <class _Func, class _InitListType, class... _Args>
+    requires is_constructible_v<decay_t<_Func>, initializer_list<_InitListType>, _Args...> && __is_callable_from<_Func>
+  _LIBCPP_HIDE_FROM_ABI explicit move_only_function(
+      in_place_type_t<_Func>, initializer_list<_InitListType> __il, _Args&&... __args) {
+    __construct<_Func>(__il, std::forward<_Args>(__args)...);
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(move_only_function&& __other) noexcept {
+    move_only_function(std::move(__other)).swap(*this);
+    return *this;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(nullptr_t) noexcept {
+    __reset();
+    return *this;
+  }
+
+  // TODO: Do we want to make this `noexcept` as an extensions?
+  template <class _Func>
+  _LIBCPP_HIDE_FROM_ABI move_only_function& operator=(_Func&& __func) noexcept {
+    move_only_function(std::forward<_Func>(__func)).swap(*this);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI ~move_only_function() { __reset(); }
+
+  // [func.wrap.move.inv]
+  _LIBCPP_HIDE_FROM_ABI explicit operator bool() const noexcept {
+    return __storage_.__large_.__status_ != _Store::_Status::_NotEngaged;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI _ReturnT operator()(_ArgTypes... __args) _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF
+      noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
+    auto __call = [&](auto&& __callable, auto&& __data) {
+      return std::bit_cast<void (*)(_LIBCPP_MOVE_ONLY_FUNCTION_CV void*, _ArgTypes...)>(__callable)(
+          __data, std::forward<_ArgTypes>(__args)...);
+    };
+
+    using enum _Store::_Status;
+    switch (__storage_.__large_.__status_) {
+    case _NotEngaged:
+      _LIBCPP_ASSERT(false, "Tried to call move_only_function which doesn't hold a callable object");
```

Comment:
Nit: `"Tried to call a disengaged move_only_function"` might be better, and is shorter. (Or just remove the word "callable"? A move_only_function physically cannot hold a _non_-callable object, so the emphasis on what kind of object it holds seems misplaced.)